modules/drivers/audio/respeaker.cc

```diff
@@ -0,0 +1,156 @@
+
+#include "modules/drivers/audio/respeaker.h"
+
+namespace apollo {
+namespace drivers {
+namespace audio {
+
+// Helper functions
+void report_error(PaError err, const std::string &func_name) {
+  AERROR << "an error occured while calling " << func_name;
+  AERROR << "error number: " << err;
+  AERROR << "error message: " << Pa_GetErrorText(err);
+  throw std::runtime_error("");
+}
+
+// Stream
+Stream::~Stream() { Pa_CloseStream(pastream_ptr_); }
+
+void Stream::init_stream(int rate, int channels, int chunk,
+                         int input_device_index, PaSampleFormat format) {
+  PaError err;
+
+  // Init parameters of input device
+  inputParameters_ptr_ = reinterpret_cast<PaStreamParameters *>(
+      malloc(sizeof(PaStreamParameters)));
+  inputParameters_ptr_->device = input_device_index;
+  inputParameters_ptr_->channelCount = channels;
+  inputParameters_ptr_->sampleFormat = format;
+  inputParameters_ptr_->suggestedLatency =
+      Pa_GetDeviceInfo(inputParameters_ptr_->device)->defaultLowInputLatency;
+  inputParameters_ptr_->hostApiSpecificStreamInfo = nullptr;
+
+  err = Pa_OpenStream(
+      &pastream_ptr_, inputParameters_ptr_, nullptr, rate, chunk,
+      paClipOff,  // we only use input so don't bother clipping them *
+      nullptr, nullptr);
+  if (err != paNoError) report_error(err, "Pa_OpenStream");
+  err = Pa_StartStream(pastream_ptr_);
+  if (err != paNoError) report_error(err, "Pa_StartStream");
+}
+
+void Stream::read_stream(int n_frames, char *buffer) {
+  int err =
+      Pa_ReadStream(pastream_ptr_, reinterpret_cast<void *>(buffer), n_frames);
+  if (err != paNoError) {
+    report_error(err, "Pa_ReadStream");
+  }
+}
+
+int Stream::get_chunk_size(int n_frames) {
+  return (n_frames) * (inputParameters_ptr_->channelCount) *
+         (Pa_GetSampleSize(inputParameters_ptr_->sampleFormat));
+}
+
+// Respeaker
+Respeaker::~Respeaker() { Pa_Terminate(); }
+void Respeaker::init(std::shared_ptr<SpeakerConfig> speaker_config) {
+  if (speaker_config->speaker_model() != SpeakerConfig::RESPEAKER) {
+    AERROR << "respeaker driver only supports respeaker model in config file";
+  }
+  int err = Pa_Initialize();
+  if (err != paNoError) {
+    Pa_Terminate();
+    report_error(err, "Pa_Initialize");
+  }
+
+  PaDeviceIndex device_index = get_respeaker_index();
+  stream_ptr_.reset(new Stream());
+  stream_ptr_->init_stream(
+      speaker_config->sample_rate(), speaker_config->channel_type_size(),
+      speaker_config->chunk(), device_index,
+      get_format_from_width(speaker_config->sample_width()));
+}
+
+PaSampleFormat Respeaker::get_format_from_width(int width, bool is_unsigned) {
+  switch (width) {
+    case 1:
+      return is_unsigned ? paUInt8 : paInt8;
+    case 2:
+      return paInt16;
+    case 3:
+      return paInt24;
+    case 4:
+      return paFloat32;
+  }
+  AERROR << "invalid width: " << width;
+  throw std::runtime_error("");
```

Comment:
Does we prefer returning error code or throwing exception for error handling @xiaoxq @Capri2014 @jinghaomiao ?

Reply:
> Does we prefer returning error code or throwing exception for error handling @xiaoxq @Capri2014 @jinghaomiao ?

No, just return of error code, we can leave this as TODO in this pr. @Crysple

Reply:
According to my experience, AERROR will block in anywhere except the block inside `while (!cyber::IsShutdown())`, which is in the `void run()` of the `component` class -- It will keep getting the same error again & again -- (since its return type is `void` I cannot return). That's the reason I used `throw`.  For elsewhere (init, etc.), I think I can change the code easily if returning error code is preferred.

Reply:
Yes, please remove the throw exception logic in this file and another

Reply:
> Yes, please remove the throw exception logic in this file and another

Ok now I remove all exceptions in related to Init. I remain the one in `read_stream` but I also add a `try catch` in `run` -- so that I can catch and break from `run` returning to apollo once there's any error.